### PR TITLE
installation script

### PR DESCRIPTION
**Pull request by JanZaion**
Hello @walmik, so I wrote the script.

Please do not feel pressured to review this PR any time sooner, than your life priorities permit, I am in no hurry with this :)

My intention here is to make the installation process similar to `npx create-react-app`, the command being `generate-scribble4max-device device-name`. All that the script does is that it copies all the contents of the 'https://github.com/JanZaion/scribbleformax' repo and then it installs all the packages. What will be the final link for downloads and what would be the contents to download (I mean do you want to include Chord and Riff?) is of course your decision.

I have tested it on my Windows machine with the `npm link` method, so I hope it works.

If you find this method suitable, it should be published as an npm package, so it works just like `npx create-react-app`.

I hope this makes sense, do not hesitate to ask me for clarification.

### CODE REVIEW

**Comment by walmik** [comment]
I think this looks great @JanZaion (thank you so much for already getting this started!)

I like the idea for publishing a NPM module for this (we can come up with a slightly smaller / more memorable name for it).

For your question of shud riff or chord be included, i think it shouldnt be included. Also these 2 devices should be re-implemented with whatever you re putting together. Basically they should be built from the script you re working on in this PR.

One small suggestion i had was for this line [here](https://github.com/scribbletune/scribbleformax/blob/master/dist/nodeScribbletune.js#L9-L15) where i was thinking we should bring in all the properties unconditionally (Scribbletune should know how to ignore props that it doesnt recognize) 

Basically it can do something like
```
const parameters = await maxApi.getDict("parameters");
const clip = scribble.clip(parameters);
```

This way other properties can come in too if the device passes them. And maybe we can rename `constructChords` to `constructClip` instead. Hope this makes sense!

**Comment by JanZaion** [comment]
@walmik all great points, I will address them sometimes in the following days

**Comment by JanZaion** [comment]
So heres my take on your points @walmik:

> we can come up with a slightly smaller / more memorable name for it

Absolutely, as the saying goes, there are only 2 hard things in programming :) Im thinking make-scribble4max

>For your question of shud riff or chord be included, i think it shouldnt be included

Agreed, including referential devices is the "Max" way of doing things after all.

>One small suggestion i had was for this line here i was thinking we should bring in all the properties unconditionally

Well, that would kinda defeat the whole point of what Im trying to do here. However, this point made me realise why my initial commit was so confusing. I want to encourage developers to write their own compositional logic before passing the parameters to Live. Riff is a great example of this, more on that later. My solution is the [following](https://github.com/scribbletune/scribbleformax/pull/10/commits/cc0052de167c020d755477ca72560856becf40cb), but there is probably still room for improvement.

>maybe we can rename constructChords to constructClip

Of course

>Also these 2 devices should be re-implemented with whatever you re putting together.

Actually I know Riff pretty well, because I have used it to learn how to work with Scribble4max. I also sometimes use it to, well riff :) So I took the liberty of rewriting Riff as you have suggested. Since I know the device, it only took me about an hour or so.

Couple of points:

- there may be some bugs, I havet tested it fully yet, but the main intent should definitely be there. Ill test it more once I get some studio time
- the main .js file is now called newriff. It had to be called something else than riff. We should be able to change it back to riff.js later
- its in a development mode now (unfrozen), so it doesent work out of the box, it requires the newriff.js file and the npm packages to be present
- I think that the repeat method that you use in Riff is a Lodash method. I have imported Lodash as an npm package to showcase one of the strengths of my solution

I hope that my rewrite of Riff will shed some light on my original intentions :)

File: src/newriff.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
+const maxApi = require('max-api');
+const scribble = require('scribbletune');
+const _ = require('lodash');
```
Comment (walmik) [inline src/newriff.js]: Can we just use a polyfill for `repeat` coz requiring the entire library like this will inject it in the resulting JS file bloating the file size.

File: src/newriff.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+maxApi.post('hello from v4 in riff');
+
+maxApi.addHandler('makeClip', () => {
+  //we create an event handler function to contain the script. For more information on how this works, see Node for Max documantation https://docs.cycling74.com/nodeformax/api/
+  const constructClip = (async () => {
+    const parameters = await maxApi.getDict('parameters'); //we fetch the main dictionary "parameters"
```
Comment (walmik) [inline src/newriff.js:17]: Love this way of getting the params from the dict instead of splitting a string. I feel silly for doing that :P

File: bin/generate-scribble4max-device.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    process.chdir(projectPath);
+
+    console.log('Installing dependencies...');
+    execSync('npm install');
+    
+    console.log('Removing useless files');
+    execSync('npx rimraf ./.git');
+    fs.rmdirSync(path.join(projectPath, 'bin'), { recursive: true });
+
+    console.log(`Installation is complete, good luck making ${projectName}!`);
```
Comment (walmik) [inline bin/generate-scribble4max-device.js:42]: Love the style of communication while setting up the device :)

File: src/nodeScribbletune.js [hunk 4]
```diff
@@ -3,11 +3,11 @@ const scribble = require('scribbletune');
 
 maxApi.addHandler("makeClip", () => { //we create an event handler function to contain the script. For more information on how this works, see Node for Max documantation https://docs.cycling74.com/nodeformax/api/
     
-    const constructChords = (async () => {
+    const constructClip = (async () => {
         const parameters = await maxApi.getDict("parameters") //we fetch the main dictionary "parameters"
         
-        const notes = parameters.notes //we retrieve the parameters
-        const pattern = parameters.pattern
+        const { notes, pattern } = parameters //we retrieve the parameters
```
Comment (walmik) [inline src/nodeScribbletune.js:9]: Are only these 2 params used here by default as they are "required" in Scribbletune and the rest are optional? However my 2 cents are that we should bring in all the params of Scribbletune in here by default and let users remove it if they dont want it. This way they can experiment adding newer features to their device using Scribbletune::clip's API without having to worry if a particular param is being imported or not.

I m yet to clone your branch and try it out hence need to quickly ask you if my understanding is correct on the following:
1. src/nodeScribbletune.js will be downloaded to the user's local machine by the "generate" script 
2. users can manipulate it further if they want 
3. users must run some script to produce/update the amxd file

**Review by walmik — COMMENTED** [review]
Hello @JanZaion I ve left a few comments (and a question) Other than that, this is really great stuff buddy. I hadnt imagined scribbleformax could be extended like this :)

File: src/newriff.js (hunk 1 above)
Comment (JanZaion) [reply src/newriff.js]: Sure, I just wanted to show its possible to do it this way :P

File: src/newriff.js (hunk 2 above)
Comment (JanZaion) [reply src/newriff.js:17]: Well hey, it took me forever to figure this out. The Dict object is pretty neat actually

File: bin/generate-scribble4max-device.js (hunk 3 above)
Comment (JanZaion) [reply bin/generate-scribble4max-device.js:42]: Well I better then give credit where credit is due, since Ive mainly followed steps in this [article](https://dev.to/leopold/generate-your-web-app-boilerplate-like-create-react-app-does-301p)

File: src/nodeScribbletune.js (hunk 4 above)
Comment (JanZaion) [reply src/nodeScribbletune.js:9]: To your point on bringing all the params in, do you mean to enrich the UI of the template device with stuff like subdiv, sizzle etc? I think thats probably a good idea.

Comment (JanZaion) [reply src/nodeScribbletune.js:9]: About the installation:

1. thats the idea
2. yes
3. Im affraid this is the biggest potential downside of this whole thing

To make the device ready for distribution, users have to jump through a few hoops, there is a detailed guide [here](https://cycling74.com/forums/node-for-max-bundling-in-collectives-standalones-and-max-for-live-devices). Heres the bad news: only users with full Max licence can do this. That means people who have Max for Live, that is Max licence obtained via purchasing Live Suite licence cannot export so called [collectives](https://docs.cycling74.com/max8/vignettes/collective_guide). Freezing the device works with js obcets, but Im 90% sure it doesent work with npm packages.

So users who own only the M4L licence can either
1. use webpack or some other bundler and Im not sure that will work 100%
2. include the `npm script install` button for first time users and make the end user install all the npm stuff on their machine

In summary, those who download it and want to use it only for their personal use dont need to worry about step 3. Those who want to give or sell devices to other producers need to have full Max license or use Webpack.

**Comment by JanZaion** [comment]
Well Im glad you like it @walmik :) Scribbletune is rightly a very popular repo, but Scribble4max deserves way more love. Too bad the "production" stage of the final amxd is not super smooth.

File: src/nodeScribbletune.js (hunk 4 above)
Comment (walmik) [reply src/nodeScribbletune.js:9]: `To make the device ready for distribution, users have to jump through a few hoops` <- this is perfectly fine Since they are already using npm to install, we can assume they are running node locally and we can provide another script that exports the module for distribution.

**Review by walmik — APPROVED** [review]

**Comment by walmik** [comment]
Thank you @JanZaion and apologies for the delay in merging this!

**Comment by JanZaion** [comment]
Hey @walmik Im glad you liked the fork. Im still planning to extend the template so that there are all the Scribbletune parameters present like we have discussed earlier here in this conversation. It should be much smaller PR anyway.

**Comment by walmik** [comment]
Hello @JanZaion I was trying out few of the newly introduced things by you so that I could push the update riff device to maxforlive.com and encountered some issues. 
1. When I drag and drop the Riff.amxd on one of the MIDI channels in Live, I get this message in Max:
```
node.script: newriff.js: No such file or directory
```

2. Once I rename this file in the webpack config and make sure that newriff is indeed used, I get the issue,
```
Module not found: Error: Can't resolve 'max-api' in '/Users/wdeshpande/Github/scribbleformax/src'
```

3. If I install this module, I get the error in the node.script debug module that suggests I shouldnt npm install "max-api" and back to step 2 :/ 

 I m sure I m missing something, could you please guide? Many thanks!

**Comment by JanZaion** [comment]
Hello @walmik, apologies for the inconvenience. I was afraid something like this might happen, since its pretty hard for me to set up "clean" environment for testing.

Do your File Preferences point correctly to the directory where newriff.js resides? If you have it set up correctly, I will look into it further

**Comment by walmik** [comment]
I think I got it @JanZaion :)
Basically the newriff.js doesnt need to go through webpack at all. It can be directly used as is. I was trying to send newriff.js from src/ to dist/ via webpack via tanspilation :P 

I think I ll spruce up some documentation and create a new package to publish to maxforlive.com. Thank you once again for introducing max-api to scribbleformax!

**Comment by JanZaion** [comment]
Yeah, I think you got it @walmik :)

> I think I ll spruce up some documentation

Feel free to let me know if you need any help with that.

> create a new package to publish to maxforlive.com.

Well... Im not sure we have addressed the problem of production (or freezing or however its called correctly) properly. 

When I was addressing this point earlier, you wrote 
> we can assume they are running node locally and we can provide another script that exports the module for distribution

I dont fully understand what did you mean by that. Did you mean writing a script for Webpack? Because Im almost certain you need to own the full Max license to freeze anything properly with Max's node script object, with or without transpilation. And yeah, there are more Max licences, its a bit confusing. If you own Ableton Live Suite, then you own Max for Live. Max for Live is not the full licence.

The upgrade to the full license costs $200 I think. So if you dont have one yourself and dont want to upgrade, let me know, I can produce Riff with my license.

...yeah, I know, the finish line is always just a few more steps away :)
